cpp/include/raft/core/serialization.hpp

```diff
@@ -0,0 +1,127 @@
+ *   when `nullptr`, the actual data is not written (only the size-to-written is calculated).
+ * @param obj the object to be serialized
+ * @param args context required for serialization
+ * @return the number of bytes (to be) written by the pointer
+ */
+template <typename T, typename... ContextArgs>
+auto serialize(uint8_t* out, const T& obj, ContextArgs&&... args) -> size_t
+{
+  return detail::call_serialize<T, ContextArgs...>(out, obj, std::forward<ContextArgs>(args)...);
+}
+
+/**
+ * @brief Write a serializable state of an object to a host vector.
+ *
+ * @tparam T type of the serializable object
+ * @tparam ContextArgs types of context required for serialization
+ *
+ * @param obj the object to be serialized
+ * @param args context required for serialization
+ * @return the serialized state
+ */
+template <typename T, typename... ContextArgs>
+auto serialize(const T& obj, ContextArgs&&... args) -> std::vector<uint8_t>
+{
+  return detail::call_serialize<T, ContextArgs...>(obj, std::forward<ContextArgs>(args)...);
+}
+
+/**
+ * @brief Read a serializable state of an object from memory.
+ *
+ * @tparam T type of the serializable object
+ * @tparam ContextArgs types of context required for serialization
+ *
+ * @param[out] p an unitialized host pointer to a location where the object should be created.
+ * @param[in] in a host pointer to the location where the state should be read from.
+ * @param args context required for serialization
+ * @return the number of bytes read by the pointer
+ */
+template <typename T, typename... ContextArgs>
+auto deserialize(T* p, const void* in, ContextArgs&&... args) -> size_t
+{
+  return detail::call_deserialize<T, ContextArgs...>(
+    p, reinterpret_cast<const uint8_t*>(in), std::forward<ContextArgs>(args)...);
+}
+
+/**
+ * @brief Read a serializable state of an object from a vector.
+ *
+ * @tparam T type of the serializable object
+ * @tparam ContextArgs types of context required for serialization
+ *
+ * @param[out] p an unitialized host pointer to a location where the object should be created.
+ * @param in a vector where the state should be read from.
+ * @param args context required for serialization
+ * @return the number of bytes read from the vector
+ */
+template <typename T, typename... ContextArgs>
+auto deserialize(T* p, const std::vector<uint8_t>& in, ContextArgs&&... args) -> size_t
+{
+  return detail::call_deserialize<T, ContextArgs...>(p, in, std::forward<ContextArgs>(args)...);
+}
+
+/**
+ * @brief Read a serializable state of an object.
+ *
+ * @tparam T type of the serializable object
+ * @tparam ContextArgs types of context required for serialization
+ *
+ * @param[in] in a host pointer to the location where the state should be read from.
+ * @param args context required for serialization
+ * @return the deserialized object;
+ */
+template <typename T, typename... ContextArgs>
+auto deserialize(const void* in, ContextArgs&&... args) -> T
```

Comment:
Instead of taking the input buffer as `const void*`, would you consider taking it as `std::span<uint8_t>`?  That would better express both the length of the buffer, and its desired element type.  Otherwise, users have no idea what input representation you intend.

Reply:
Let's keep the related discussion here
https://github.com/rapidsai/raft/pull/770#discussion_r935679007